### ISSUE TITLE
Venom library releaseImplementation vs debugImplementation

### ISSUE THREAD

**Issue by ColtonIdle** [autonomousapps/dependency-analysis-android-gradle-plugin#340]
**Plugin version**
~Latest~ 0.68.0

**Gradle version**
6.7

**Android Gradle Plugin (AGP) version**
4.1.1

**Describe the bug**
Added this dependency https://github.com/YarikSOffice/venom and now my scans say

```
Existing dependencies which should be modified to be as indicated:
- implementation("com.github.YarikSOffice.Venom:venom-no-op:0.3.1") (was releaseImplementation)
- implementation("com.github.YarikSOffice.Venom:venom:0.3.1") (was debugImplementation)
```

but that seems right. I should use no-op for release and actual impl for debug.

**Comment by autonomousapps** [comment]
Thanks for the issue.

First, please do include the actual plugin version when filing issues. I release often enough that if an issue is even a week old, I'll have a hard time knowing what "latest" refers to.

That said, this definitely looks like a bug, and I thought I had this case handled already. Will take a look.

**Comment by ColtonIdle** [comment]
You got it! Made the change and will remember for the future.

**Comment by autonomousapps** [comment]
for now, you ought to be able to work around this via https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/wiki/Customizing-plugin-behavior#failure-conditions-and-filtering

**Comment by autonomousapps** [comment]
This is a complicated case to resolve, because from gradle's perspective, the "real" and "no-op" versions of this library are actually just two separate, unrelated libraries. I can imagine a heuristic that looks at all the advice and decides that two pieces of advice are conflicting because the dependencies provide, in some sense, the same "capability"; but without the libraries actually using gradle's Capabilities API, I'd have to do a potentially very expensive comparison of all dependencies to all others and see if they provide some of the same classes (for example). And that wouldn't capture the case where the libraries don't provide classes but only resources or native libraries or etc etc.

I think for now your best case is to exclude these libraries as described [here](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/wiki/Customizing-plugin-behavior#failure-conditions-and-filtering) or, alternatively, make use of your debug source set and create a debug `Application` class and only have _it_ use this library (the "real" version). I have confirmed that if you do this, the plugin no longer gives conflicting advice.

**Comment by ColtonIdle** [comment]
I will exclude the libraries! Just wanted to bring it up to see if there was any other underlying issue.

I'm curious if this happens to a lot of other no-op libraries that I've used?

e.g. FB Flipper debug tool
 debugImplementation 'com.facebook.flipper:flipper:0.68.0'
 releaseImplementation 'com.facebook.flipper:flipper-noop:0.68.0'

Chucker network debug tool
  debugImplementation "com.github.chuckerteam.chucker:library:3.4.0"
  releaseImplementation "com.github.chuckerteam.chucker:library-no-op:3.4.0"

**Comment by autonomousapps** [comment]
It may very well happen with other libraries that follow this pattern, but I can't say for sure in each case without testing. Android libraries in particular have a lot of potential capabilities that are all analyzed separately and make various contributions to the final advice.

**Comment by autonomousapps** [comment]
For what it's worth, my position is that this real/noop pattern is really an anti-pattern. I hope it's a mere historical legacy that we can (sooner than later) evolve out of. You can already use maven-publish to publish a variant-aware library that has different behavior for debug vs release, all published to the same maven coordinates.

**Comment by ColtonIdle** [comment]
makes sense. just wanted to see if I was the one at fault here.

noted about the anti-pattern. I don't know how I feel about it, but it does make it easy to drop in a debug only dependency.

I might follow your lead on this "alternatively, make use of your debug source set and create a debug Application class and only have it use this library (the "real" version). I have confirmed that if you do this, the plugin no longer gives conflicting advice."

**Comment by autonomousapps** [comment]
You're not at fault. It's a real pattern we see in the Android community. And it may just be my opinion that it's an anti-pattern.

**Comment by autonomousapps** [comment]
This issue has been fixed with the linked PR. Will be released as part of 0.70.0.
